feat: add support for charging devices to measure tool

bramstroker opened the pull request.

bramstroker commented:
@RubenKelevra I did some more work on this today, and first version should be finished now.
Are you able to test this? By checking out the branch `feat/measure-util-charging-device`.

```
cd utils/measure
poetry install --no-root
poetry run python measure.py
```

RubenKelevra commented:
> Are you able to test this?

Sure!

RubenKelevra commented:
This feels a bit redundant, maybe it's possible to skip this step if there's a `battery_level` attribute in the list?

``` 
[?] Select the battery_level attribute: 
   battery_icon
 > battery_level
   fan_speed
   fan_speed_list
   friendly_name
   icon
   supported_features
```

RubenKelevra commented:
Got that crash while the vacuum was on the way back to his dock, and I enter.

```
Make sure the device is as close to 0% charged as possible before starting the test.
Hit enter when you are ready to start..
Traceback (most recent call last):
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/measure.py", line 335, in <module>
    main()
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/measure.py", line 327, in main
    measure.start()
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/measure.py", line 135, in start
    runner_result = self.runner.run(answers, export_directory)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/runner/charging.py", line 51, in run
    is_charging = self.controller.is_charging()
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/controller/charging/hass.py", line 39, in is_charging
    entity = self.client.get_entity(self.entity_id)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ruben/.cache/pypoetry/virtualenvs/measure-DQAfURof-py3.12/lib/python3.12/site-packages/homeassistant_api/rawclient.py", line 230, in get_entity
    raise ValueError(
ValueError: Neither group_id and slug or entity_id provided. Use keyword arguments to pass entity_id. Or you can pass the group_id and slug instead


```

RubenKelevra commented:
And the same once the device started charging:


```
Make sure the device is as close to 0% charged as possible before starting the test.
Hit enter when you are ready to start..
Traceback (most recent call last):
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/measure.py", line 335, in <module>
    main()
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/measure.py", line 327, in main
    measure.start()
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/measure.py", line 135, in start
    runner_result = self.runner.run(answers, export_directory)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/runner/charging.py", line 51, in run
    is_charging = self.controller.is_charging()
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/controller/charging/hass.py", line 39, in is_charging
    entity = self.client.get_entity(self.entity_id)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ruben/.cache/pypoetry/virtualenvs/measure-DQAfURof-py3.12/lib/python3.12/site-packages/homeassistant_api/rawclient.py", line 230, in get_entity
    raise ValueError(
ValueError: Neither group_id and slug or entity_id provided. Use keyword arguments to pass entity_id. Or you can pass the group_id and slug instead

```

bramstroker commented:
> This feels a bit redundant, maybe it's possible to skip this step if there's a `battery_level` attribute in the list?

Good idea, will add this.

bramstroker commented:
> Got that crash while the vacuum was on the way back to his dock, and I enter.

I see, made a mistake in the hass controller code. Did only do my latest tests with the `dummy` controller as I don't have a vacuum entity myself.
Hopefully should be fixed now.

bramstroker commented:
> > This feels a bit redundant, maybe it's possible to skip this step if there's a `battery_level` attribute in the list?
> 
> Good idea, will add this.

This should also be implemented now.

RubenKelevra commented:
> > Got that crash while the vacuum was on the way back to his dock, and I enter.
> 
> I see, made a mistake in the hass controller code. Did only do my latest tests with the `dummy` controller as I don't have a vacuum entity myself. Hopefully should be fixed now.

Still crashes, now at a different location:

``` 
❯ poetry run python measure.py
Powercalc measure: master

2024-11-03 21:10:11,779 [INFO] Selected powermeter: shelly
2024-11-03 21:10:11,780 [INFO] Selected light controller: hass
[?] What kind of measurement session do you want to run?: 
   Light bulb(s)
   Smart speaker
   Recorder
   Average
 > Charging device

[?] Specify the full Charging device model name: Roborock S6 MaxV
[?] Select the charging device type: 
 > vacuum robot

[?] Select the vacuum entity: 
 > vacuum.herbie

Make sure the device is as close to 0% charged as possible before starting the test.
Hit enter when you are ready to start..
Traceback (most recent call last):
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/measure.py", line 335, in <module>
    main()
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/measure.py", line 327, in main
    measure.start()
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/measure.py", line 135, in start
    runner_result = self.runner.run(answers, export_directory)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/runner/charging.py", line 46, in run
    battery_level = self.controller.get_battery_level()
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/controller/charging/hass.py", line 36, in get_battery_level
    return int(entity.state.attributes[self.battery_level_attribute])
               ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
KeyError: None
```

RubenKelevra commented:
So your is_charging should be modified and look like this:

```
39     def is_charging(self) -> bool:
40         """Check if the device is currently charging"""
41
42         entity = self.client.get_entity(entity_id=self.entity_id)
43         return (entity.state.state == "docked")
```

Because entity.state looks like this:

``` 
entity_id='vacuum.herbie' state='idle' attributes={'fan_speed_list': ['off', 'quiet', 'balanced', 'turbo', 'max', 'custom', 'max_plus'], 'battery_level': 18, 'battery_icon': 'mdi:battery-20', 'fan_speed': 'custom', 'icon': 'mdi:robot-vacuum', 'friendly_name': 'Herbie', 'supported_features': 14204} last_changed=datetime.datetime(2024, 11, 3, 20, 11, 49, 62926, tzinfo=TzInfo(UTC)) last_updated=datetime.datetime(2024, 11, 3, 20, 11, 49, 62926, tzinfo=TzInfo(UTC)) context=Context(id='XXXXXXX')
``` 

And the bug mentioned before seems to stem from some issue determining the string to find the battery level. So get_battery_level() works fine with this return hacked in:

```
return int(entity.state.attributes["battery_level"])
```

RubenKelevra commented:
I also recommend not raising an error, and instead just looping, if not (yet) is_charging() is true. Say if it's `idle` or `returning`. Currently this would result in this trace being thrown at the user:

``` 
Traceback (most recent call last):
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/measure.py", line 335, in <module>
    main()
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/measure.py", line 327, in main
    measure.start()
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/measure.py", line 135, in start
    runner_result = self.runner.run(answers, export_directory)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/runner/charging.py", line 53, in run
    raise RunnerError("Device is not charging anymore.")
runner.errors.RunnerError: Device is not charging anymore.

``` 

I think "waiting for vacuum cleaner to start charging... current status: idle" or something like this, would be a bit more userfriendly ;)

RubenKelevra commented:
Ha... looks like the battery percentage dropped quite a bit and the robot did not update, until charging started:

``` 
2024-11-03 22:30:52,078 [INFO] Battery level: 18%
2024-11-03 22:30:58,138 [INFO] Measured power: 7.96 W
2024-11-03 22:31:01,151 [INFO] Battery level: 18%
2024-11-03 22:31:07,207 [INFO] Measured power: 7.99 W
2024-11-03 22:31:10,222 [INFO] Battery level: 18%
2024-11-03 22:31:16,278 [INFO] Measured power: 8.05 W
2024-11-03 22:31:19,308 [INFO] Battery level: 18%
2024-11-03 22:31:25,378 [INFO] Measured power: 7.96 W
2024-11-03 22:31:28,393 [INFO] Battery level: 18%
2024-11-03 22:31:34,467 [INFO] Measured power: 8.03 W
2024-11-03 22:31:37,486 [INFO] Battery level: 18%
2024-11-03 22:31:43,558 [INFO] Measured power: 7.96 W
2024-11-03 22:31:46,574 [INFO] Battery level: 18%
2024-11-03 22:31:52,650 [INFO] Measured power: 7.92 W
2024-11-03 22:31:55,669 [INFO] Battery level: 18%
2024-11-03 22:32:01,725 [INFO] Measured power: 7.97 W
2024-11-03 22:32:04,738 [INFO] Battery level: 18%
2024-11-03 22:32:10,798 [INFO] Measured power: 8.04 W
2024-11-03 22:32:13,814 [INFO] Battery level: 18%
2024-11-03 22:32:19,869 [INFO] Measured power: 8.08 W
2024-11-03 22:32:22,879 [INFO] Battery level: 18%
2024-11-03 22:32:28,942 [INFO] Measured power: 8.01 W
2024-11-03 22:32:31,956 [INFO] Battery level: 9%
2024-11-03 22:32:38,025 [INFO] Measured power: 7.94 W
2024-11-03 22:32:41,053 [INFO] Battery level: 9%
2024-11-03 22:32:47,106 [INFO] Measured power: 8.02 W
2024-11-03 22:32:50,119 [INFO] Battery level: 9%
2024-11-03 22:32:56,181 [INFO] Measured power: 8.00 W
2024-11-03 22:32:59,201 [INFO] Battery level: 9%
2024-11-03 22:33:05,255 [INFO] Measured power: 7.98 W
2024-11-03 22:33:08,270 [INFO] Battery level: 9%
2024-11-03 22:33:14,326 [INFO] Measured power: 7.92 W
2024-11-03 22:33:17,342 [INFO] Battery level: 9%

```  

Another edge case for you right there ;)

RubenKelevra commented:
Reading worked until the Shelly power meter reported a 0 - for whatever reason 🤔

``` 
2024-11-04 02:05:23,859 [INFO] Battery level: 87%
2024-11-04 02:05:29,986 [INFO] Measured power: 6.74 W
2024-11-04 02:05:33,000 [INFO] Battery level: 87%
2024-11-04 02:05:39,056 [INFO] Measured power: 6.66 W
2024-11-04 02:05:42,080 [INFO] Battery level: 87%
2024-11-04 02:05:48,139 [INFO] Measured power: 6.69 W
2024-11-04 02:05:51,166 [INFO] Battery level: 87%
2024-11-04 02:05:57,239 [INFO] Measured power: 6.66 W
2024-11-04 02:06:00,253 [INFO] Battery level: 88%
2024-11-04 02:06:06,314 [INFO] Measured power: 6.63 W
2024-11-04 02:06:09,332 [INFO] Battery level: 88%
2024-11-04 02:06:15,385 [INFO] Measured power: 6.67 W
2024-11-04 02:06:18,405 [INFO] Battery level: 88%
2024-11-04 02:06:24,548 [INFO] Measured power: 6.59 W
2024-11-04 02:06:27,562 [INFO] Battery level: 88%
2024-11-04 02:06:42,749 [ERROR] Aborting: 0 watt was read from the power meter

```

RubenKelevra commented:
Looks like the drop to 0 W is normal and the trickle charge/standby is started a bit later once a certain voltage is reached by drawing from the battery.

The blip of charging later seems not normal. 

Here's my second run:

![Screenshot_2024-11-04-17-27-41-925-edit_io.homeassistant.companion.android.jpg](https://github.com/user-attachments/assets/547f7525-7086-4134-8f06-1586f7e3493d)

bramstroker commented:
Thanks for the first tests @RubenKelevra, much appreciated.
Will be some days probably next weekend until I can come back on it. Busy schedule this week.
Also optimizations / PR for the code is welcome.
You can just branch of my branch and than create a PR and target to feat/measure-util-charging-device, than we can incoorporate the changes for this feature.

RubenKelevra commented:
Alright, on the second look it just looks like the charging just starts trickling in with some weird power pattern or so low power, that the Shelly Plug doesn't pick it up:

![Screenshot_20241104_225828](https://github.com/user-attachments/assets/db84b1df-7146-4792-a601-08e4b1a3e8c0)

Means, I need a dummy load this time again, which I didn't used because of space restrictions in the area the robot sits (under a cabinet).

Anyway, I set it up again, and try again. :)

RubenKelevra commented:
Here are my fixes:

https://github.com/bramstroker/homeassistant-powercalc/pull/2643

RubenKelevra commented:
Btw, there's also a difficulty in this for the users which we may want to document:

Many robot vacuums these days will automatically return to their base to recharge a bit and then resume cleaning when they hit low battery. This can lead them to continue at around 80% SoC, which isn't ideal for our measurement purposes.

For Roborock models, there’s a workaround using the "Pin to Go" function—users can send the robot to a specific area and start a spot clean, even if the SoC is below 20%. This interrupts the cleaning job, so the robot won’t try to resume an unfinished task afterward. This approach helps us to drain the battery a bit past 20% and also avoids that the robot continues on a cleaning job on its own.

Also, there’s no need to discharge down to 0% SoC, as currently instructed. Most vacuums will start heading back to the dock around 20%, only consuming a small amount (about 1-2% SoC) on their way back. So it’s enough to run them slightly below the self-return threshold before starting the measurement process, as in real life applications they won't go down to 0%.

RubenKelevra commented:
With the fixes and a dummy load, it now works fine...

```
❯ poetry run python measure.py
Powercalc measure: master

2024-11-05 00:57:03,538 [INFO] Selected powermeter: shelly
2024-11-05 00:57:03,539 [INFO] Selected light controller: hass
[?] What kind of measurement session do you want to run?: 
   Light bulb(s)
   Smart speaker
   Recorder
   Average
 > Charging device

[?] Specify the full Charging device model name: Roborock S6 MaxV
[?] Select the charging device type: 
 > vacuum robot

[?] Select the vacuum entity: 
 > vacuum.herbie

Make sure the device is as close to 0% charged as possible before starting the test.
Hit enter when you are ready to start..

waiting for vacuum cleaner to start charging...
vacuum cleaner started charging, starting measurements
2024-11-05 00:59:24,591 [INFO] Battery level: 15%
2024-11-05 00:59:30,647 [INFO] Measured power: 16.48 W
2024-11-05 00:59:33,666 [INFO] Battery level: 15%
``` 

Except that it does not subtract the dummy load...

And looks like there was a connection hiccup for a minute and 4 seconds, which means the battery percentage couldn't be fetched and thus it stopped with a crash:

```
2024-11-05 03:11:41,312 [INFO] Measured power: 30.84 W
2024-11-05 03:11:44,335 [INFO] Battery level: 50%
2024-11-05 03:11:50,393 [INFO] Measured power: 30.92 W
2024-11-05 03:11:53,414 [INFO] Battery level: 50%
2024-11-05 03:11:59,482 [INFO] Measured power: 30.76 W
2024-11-05 03:12:02,502 [INFO] Battery level: 50%
2024-11-05 03:12:08,562 [INFO] Measured power: 31.00 W
2024-11-05 03:12:11,580 [INFO] Battery level: 50%
2024-11-05 03:12:17,646 [INFO] Measured power: 31.20 W
Traceback (most recent call last):
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/measure.py", line 335, in <module>
    main()
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/measure.py", line 327, in main
    measure.start()
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/measure.py", line 135, in start
    runner_result = self.runner.run(answers, export_directory)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/runner/charging.py", line 69, in run
    battery_level = self.controller.get_battery_level()
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/ruben/Projects/bramstroker_homeassistant-powercalc_for_light_measurements/utils/measure/controller/charging/hass.py", line 36, in get_battery_level
    return int(entity.state.attributes[self.battery_level_attribute])
               ~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
KeyError: 'battery_level'
``` 

![Screenshot_20241105_135338](https://github.com/user-attachments/assets/c70572d6-702d-46ce-95e1-4dd0622c9333)

RubenKelevra commented:
> looks like there was a connection hiccup for a minute and 4 seconds, which means the battery percentage couldn't be fetched and thus it stopped with a crash

Thought about that: I think it's fine to skip errors like this and fetch just the power meter values until the robot starts responding again.

But that would only work if the robot is reporting the same battery percentage after the interruption again - which is likely for such small hiccups in the wifi connectivity.

I'll work on something like that later today.

bramstroker commented:
> Btw, there's also a difficulty in this for the users which we may want to document:
> 
> Many robot vacuums these days will automatically return to their base to recharge a bit and then resume cleaning when they hit low battery. This can lead them to continue at around 80% SoC, which isn't ideal for our measurement purposes.
> 
> For Roborock models, there’s a workaround using the "Pin to Go" function—users can send the robot to a specific area and start a spot clean, even if the SoC is below 20%. This interrupts the cleaning job, so the robot won’t try to resume an unfinished task afterward. This approach helps us to drain the battery a bit past 20% and also avoids that the robot continues on a cleaning job on its own.
> 
> Also, there’s no need to discharge down to 0% SoC, as currently instructed. Most vacuums will start heading back to the dock around 20%, only consuming a small amount (about 1-2% SoC) on their way back. So it’s enough to run them slightly below the self-return threshold before starting the measurement process, as in real life applications they won't go down to 0%.

Might be added to https://docs.powercalc.nl/contributing/measure/
Or even separate page: https://docs.powercalc.nl/contributing/measure/vacuum

bramstroker commented:
> And looks like there was a connection hiccup for a minute and 4 seconds, which means the battery percentage couldn't be fetched and thus it stopped with a crash:

Added some better error handling with:

https://github.com/bramstroker/homeassistant-powercalc/pull/2603/commits/5abc8e7bd7711c9c402311b976550a7e9fb2be53

Should continue trying again for a few times now. Untested.

bramstroker commented:
Made it possible to use a dummy load for every measurement type, including charging with https://github.com/bramstroker/homeassistant-powercalc/pull/2603/commits/13797f403e30149dab65742564edb1e7abda4179.
Also fixed a typing issue I introduced with my latest commit.

@RubenKelevra Are you able to test again with all the improvements done?

RubenKelevra commented:
@bramstroker yeah, but only today. Yesterday I've set up another device already for testing. 🥲

RubenKelevra commented:
I'm rerunning the test, but I've increased the error count in utils/measure/runner/charging.py to 200 instead of 10, just to be sure it's now completing, as I really don't wanna run down the battery so often ;)

RubenKelevra commented:
Alright, it completed this run. :)

I think these things could be improved:

1) It now measures for 2 seconds once it reached 100% SoC the trickle charge. But we've seen other devices continue to charge at 100%. So I think it makes sense to implement my idea to continue to monitor the device for another hour after it reached 100%, and average each say 5-minute slots.

This way we're sure we measure the complete charging characteristics, as currently we would measure an extremely high trickle charge current this way for devices like the one by @KrzysztofHajdamowicz:

https://github.com/bramstroker/homeassistant-powercalc/issues/639#issuecomment-1208495005

2) Minor detail, but we write out the model.json now plain in the charging folder... Which could end up overriding a second measurement a user takes. Maybe just create a subfolder with the name the user enters or something like that?

```
2024-11-11 17:19:55,945 [INFO] Measurement session finished. Files exported to [...]/utils/measure/export/charging
```

3) We don't measure the idle consumption. My model.json shows: `"standby_power": 0` which isn't correct. The charging dock uses a bit of energy even if the robot is not *there* charging, because it runs a light beacon, so the robot can return to the base. Mine is around 0.1 W.

We could easily extend the script and measure it before we start charging the robot, as the script is now waiting for the robot to start charging.

bramstroker commented:
Oops I have set trickle time to 2 seconds for test. It was 1800.
Corrected again: https://github.com/bramstroker/homeassistant-powercalc/pull/2603/commits/b7a499aa14ec5b89efd4b1bebed74efaff433301

RubenKelevra commented:
@bramstroker yeah, doesn't change anything for my device, so doesn't affect https://github.com/bramstroker/homeassistant-powercalc/pull/2660

![Screenshot_20241111_201926](https://github.com/user-attachments/assets/c832d2f5-06f0-45cc-bacd-2ba919f21e6f)

But from this graph by @KrzysztofHajdamowicz it looks more like we need 40 instead of 30 minutes for trickle charge measurement:

https://user-images.githubusercontent.com/24268470/183493832-5aa72357-f0f5-434e-b4e5-aee4faf07436.png

And we would need a second measurement after that... so the first xx minutes and then what the device is using after trickle charge is completed. Otherwise, we would assume it's trickle charging continuously.

bramstroker commented:
@KrzysztofHajdamowicz when you are still around could you share some more state history of your vacuum? Would be nice to have some information on multiple runs, and also data for some time after the vacuum bot stopped.
Does it keep power at 0? Of does it keep trickle once in a while?

For now I'm going to merge this MR. Can improve upon it when more information is shared for your use case.
